feat: add Beta support for Inline Completion Request

rubenporras opened the pull request.

rubenporras commented:
Adds the inline completion request as proposed for LSP 3.18 (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_inlineCompletion).

I have added the Beta annotation as suggested in https://github.com/eclipse-lsp4j/lsp4j/pull/856 but this makes tests fail because this dependency is not allowed.

There are also some problems with the javadoc which I have not yet figured out. If you are interested in the PR, I would clean this up.

Are you interested in the PR? In that case, since this is my first contribution to the project, do you see any problem in how I translated the spec to the Xtend file?

Thanks

angelozerr commented:
Thanks so much @rubenporras to work on this issue. LSP4IJ (LSP support for IntelliJ) is super interested with that, we wait for this support https://github.com/redhat-developer/lsp4ij/issues/643

pisv commented:
@rubenporras Congratulations on your first contribution! It is really appreciated. I have already started a review.

pisv reviewed (changes requested):
@rubenporras I have completed the first round of review. If possible, please address review comments with a separate commit. Thank you for your time and effort!

rubenporras commented on org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:
```diff
@@ -11161,3 +11170,250 @@ class NotebookDocumentIdentifier {
+class InlineCompletionParams extends TextDocumentPositionParams implements WorkDoneProgressParams {
+  Either<String, Integer> token
+
+	/**
+	 * Additional information about the context in which inline completions
+	 * were requested.
+	 */
+	@NonNull
+	InlineCompletionContext context
```
I do not know what is off in the indentation, could you be more specific?

rubenporras commented on org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:
```diff
@@ -11161,3 +11170,250 @@ class NotebookDocumentIdentifier {
+@Beta
+@JsonRpcData
+class InlineCompletionContext {
+	/**
+	 * Describes how the inline completion was triggered.
+	 */
+	@NonNull
+	InlineCompletionTriggerKind triggerKind
+
+	/**
+	 * Provides information about the currently selected item in the
+	 * autocomplete widget if it is visible.
+	 *
+	 * If set, provided inline completions must extend the text of the
+	 * selected item and use the same range, otherwise they are not shown as
+	 * preview.
+	 * As an example, if the document text is `console.` and the selected item
+	 * is `.log` replacing the `.` in the document, the inline completion must
+	 * also replace `.` and start with `.log`, for example `.log()`.
+	 *
+	 * Inline completion providers are requested again whenever the selected
+	 * item changes.
+	 */
+	SelectedCompletionInfo selectedCompletionInfo
```
I do not know what is off in the indentation, could you be more specific?

rubenporras commented:
@pisv, thanks a lot for the thorough review.

I have addressed all the comment but the ones about indentation being a bit off because I could not figure it out. 

I have also fixed the errors reported in the javadoc when running the gradle build.

I still need advice on how to fix the problem with the Beta annotation. At the moment the annotation makes the Lsp4jArchitectureTest.testNoDependenyToGuava fail.

rubenporras commented:
I have been looking at Lsp4jArchitectureTest.testNoDependenyToGuava, and my proposal would be that to avoid the dependency we can declare our own Beta annotation in a new package org.eclipse.lsp4j.annotations. That should be straightforward.

pisv commented:
+1 for declaring our own `Beta` annotation to avoid a dependency on Guava. I think that it should be defined in `org.eclipse.lsp4j.jsonrpc` to allow it to be used everywhere in LSP4J, including `org.eclipse.lsp4j.debug`. Not sure whether we need a separate package for it. It looks like we have not had special packages for other LSP4J annotations, so I'd suggest placing it in an existing package, e.g. `org.eclipse.lsp4j.jsonrpc.util`.

pisv commented:
@rubenporras Unfortunately, I will not be able to make another round of thorough review until next week. But rest assured that your work is much appreciated! Thank you.

rubenporras commented:
Thanks for the input. 

I have added another commit on top with an own Beta annotation.

I will also not be available until next Tuesday. I guess you will enjoy the Easter time the same as I :).

rubenporras commented:
Thanks for the input. 

I have added another commit on top with an own Beta annotation.

I will also not be available until next Tuesday. I guess you will enjoy the Easter time the same as I :).

rubenporras commented on org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Beta.java:
Yes, this is correct. 

I have just removed the `@GwtCompatible` annotation and adapted the copyright to add _**and othres**_.

I could start from zero if needed, but other than having a different javadoc, there is not much room for deviating.

pisv replied:
So, the question is whether a review from the Eclipse IP Team would be required because of that and if so, whether it would be more straightforward if we could do without the elaborate Javadoc and make our own annotation from scratch instead.

pisv reviewed (commented):
@rubenporras So, I've done my best to review your changes against the spec, and they look good to me now. In particular, I have verified that the mapping of the relevant spec constructs to LSP4J looks correct. Thank you for the great work! 👍

Let us wait for feedback from @jonahgraham regarding the `Beta` annotation...

rubenporras commented:
@pisv , thanks for your review.

pisv commented:
@rubenporras Would you push the commit with the latest changes? :-) Thanks!

rubenporras commented:
> @rubenporras Would you push the commit with the latest changes? :-) Thanks!

You mean, I should rebase, squash and push?

pisv commented:
No, I mean that you have resolved the three final nitpicks I left after the latest round of review, but have not pushed the changes yet, as far as I can see :-)

rubenporras commented:
Sorry, done by now

pisv reviewed (approved):
I approve this PR from my side, because all my comments have been properly addressed, but cannot merge it until @jonahgraham has an opportunity to review the current state and submit feedback. (By way of reminder, there is an open question about the `Beta` annotation.)

rubenporras commented:
@jonahgraham ping on the feedback regarding the Beta annotation.

rubenporras commented:
@pisv , it looks like we cannot get the review from @jonahgraham. I propose that I change the javadoc comment to something more straight forward and sidestep the problem. Do you agree?

jonahgraham commented:
(Sorry for being a little delinquent here - I have been unavailable for much of the last two months)

pisv commented:
@jonahgraham Hope you are doing well, and sorry for bothering you!

Thank you for your helpful review. If I understood you correctly, the main thing about this annotation is that it needs to be implemented *from scratch* and contributed under the project's license. And, while the name of the annotation is probably not very important in this regard, you are proposing to choose a name different from the current `Beta` to reinforce that the annotation has been made specifically for using within LSP4J.

If I may still have your attention, another question about this PR or, rather, our next release in general is that would it be fine to have it merged as part of the `1.0.0` release? If I remember correctly, one of the ideas initially was that the `1.0.0` version could be released with no changes to the codebase (except the version bump itself) to set a baseline for future releases. Do you think it is still relevant? I'm asking because so far, we have not merged anything to `1.0.0`, except #889. Thanks!

jonahgraham commented:
> @jonahgraham Hope you are doing well, and sorry for bothering you!

I am (enough) - and you aren't bothering me :-)

> If I understood you correctly [...]

Yes on all accounts. Note that there is probably only one way to implement such an annotation, and that there are probably no IP issues with that one way of doing it, so I would expect that the "LSPBeta" or whatever it is called, may look a lot like guava's Beta.

> If I remember correctly, one of the ideas initially was that the 1.0.0 version could be released with no changes to the codebase (except the version bump itself) to set a baseline for future releases.

Yes that was intention.

> Do you think it is still relevant?

No, not really important.

The idea of 1.0 release is that from that point forward we are going to be API versioned according to semantic versioning.

pisv commented:
Got it. Thanks a lot for the swift response, and all your comments!

jonahgraham commented:
No objection from me to calling it Beta.

rubenporras commented:
I have renamed it to Draft, which in our case conveys better what it is

jonahgraham commented:
> I have renamed it to Draft, which in our case conveys better what it is

+1

pisv reviewed (approved):
Excellent! Thanks a lot for your contribution, @rubenporras 👍
